src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.PerAnalyzerState.cs

```diff
@@ -0,0 +1,180 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// Stores the current partial analysis state for an analyzer.
+    /// </summary>
+    internal partial class AnalysisState
+    {
+        private class PerAnalyzerState
+        {
+            private readonly ConcurrentDictionary<CompilationEvent, AnalyzerStateData> _pendingEvents = new ConcurrentDictionary<CompilationEvent, AnalyzerStateData>();
+            private readonly ConcurrentDictionary<ISymbol, AnalyzerStateData> _pendingSymbols = new ConcurrentDictionary<ISymbol, AnalyzerStateData>();
+            private readonly ConcurrentDictionary<SyntaxNode, DeclarationAnalyzerStateData> _pendingDeclarations = new ConcurrentDictionary<SyntaxNode, DeclarationAnalyzerStateData>();
+            private readonly ConcurrentDictionary<SyntaxTree, AnalyzerStateData> _pendingSyntaxAnalysisTrees = new ConcurrentDictionary<SyntaxTree, AnalyzerStateData>();
+
+            public ImmutableArray<CompilationEvent> PendingEvents => _pendingEvents.Keys.ToImmutableArray();
+
+            public PerAnalyzerState(Compilation compilation)
+            {
+                foreach (var tree in compilation.SyntaxTrees)
+                {
+                    _pendingSyntaxAnalysisTrees[tree] = null;
+                }
+            }
+
+            public bool HasPendingSyntaxAnalysis(SyntaxTree treeOpt)
+            {
+                return treeOpt != null ?
+                    _pendingSyntaxAnalysisTrees.ContainsKey(treeOpt) :
+                    _pendingSyntaxAnalysisTrees.Count > 0;
+            }
+
+            public bool HasPendingSymbolAnalysis(ISymbol symbol)
+            {
+                return _pendingSymbols.ContainsKey(symbol);
+            }
+
+            private static bool TryStartProcessingEntity<TAnalysisEntity, TAnalyzerStateData>(TAnalysisEntity analysisEntity, ConcurrentDictionary<TAnalysisEntity, TAnalyzerStateData> pendingEntities, out TAnalyzerStateData newState)
+                where TAnalyzerStateData : AnalyzerStateData, new()
+            {
+                TAnalyzerStateData currentState;
+                if (pendingEntities.TryGetValue(analysisEntity, out currentState) &&
+                    (currentState == null || currentState.StateKind == StateKind.Ready))
+                {
+                    newState = (TAnalyzerStateData)((currentState ?? new TAnalyzerStateData()).WithStateKind(StateKind.InProcess));
```

Comment:
We are checking for `currentState == null || currentState.StateKind == StateKind.Ready`. I have also removed the concurrent dictionary and wrapped this method in a lock, so should be more explicit in next iteration.